gdal/frmts/gtiff/tif_jxl.c

```diff
@@ -0,0 +1,826 @@
+        }*/
+
+        return 1;
+}
+
+/*
+* Decode a strip, tile or scanline.
+*/
+static int
+JXLDecode(TIFF* tif, uint8_t* op, tmsize_t occ, uint16_t s)
+{
+        static const char module[] = "JXLDecode";
+        JXLState* sp = DecoderState(tif);
+
+        (void) s;
+        assert(sp != NULL);
+        assert(sp->state == LSTATE_INIT_DECODE);
+
+        if( sp->uncompressed_buffer == 0 )
+        {
+            TIFFErrorExt(tif->tif_clientdata, module,
+                         "Uncompressed buffer not allocated");
+            return 0;
+        }
+
+        if( (uint64_t)sp->uncompressed_offset +
+                                        (uint64_t)occ > sp->uncompressed_size )
+        {
+            TIFFErrorExt(tif->tif_clientdata, module,
+                         "Too many bytes read");
+            return 0;
+        }
+
+        memcpy(op,
+               sp->uncompressed_buffer + sp->uncompressed_offset,
+               occ);
+        sp->uncompressed_offset += (unsigned)occ;
+
+        return 1;
+}
+
+static int
+JXLSetupEncode(TIFF* tif)
+{
+        JXLState* sp = EncoderState(tif);
+
+        assert(sp != NULL);
+        if (sp->state & LSTATE_INIT_DECODE) {
+            sp->state = 0;
+        }
+
+        sp->state |= LSTATE_INIT_ENCODE;
+
+        return 1;
+}
+
+/*
+* Reset encoding state at the start of a strip.
+*/
+static int
+JXLPreEncode(TIFF* tif, uint16_t s)
+{
+        static const char module[] = "JXLPreEncode";
+        JXLState *sp = EncoderState(tif);
+        TIFFDirectory *td = &tif->tif_dir;
+
+        (void) s;
+        assert(sp != NULL);
+        if( sp->state != LSTATE_INIT_ENCODE )
+            tif->tif_setupencode(tif);
+
+        if( td->td_planarconfig == PLANARCONFIG_CONTIG &&
+            td->td_samplesperpixel > 4 )
+        {
+            TIFFErrorExt(tif->tif_clientdata, module,
+                "JXL: Only up to 4 samples are supported for Planarconfig=Contig");
+            return 0;
+        }
+
+        if( !sp->lossless && td->td_planarconfig == PLANARCONFIG_CONTIG )
+        {
+            switch(td->td_samplesperpixel) {
+                case 2:
+                    if(td->td_sampleinfo[1] != EXTRASAMPLE_UNASSALPHA) {
+                        TIFFErrorExt(tif->tif_clientdata, module,
+                            "JXL: For 2-band images in lossy mode the second band must be alpha");
+                        return 0;
+                    }
+                    break;
+                case 3:
+                    if( td->td_photometric != PHOTOMETRIC_RGB ) {
+                        TIFFErrorExt(tif->tif_clientdata, module,
+                            "JXL: For 3-band images in lossy mode bands must be RGB");
+                        return 0;
+                    }
+                    break;
+                case 4:
+                    if( td->td_photometric != PHOTOMETRIC_RGB ||
+                        td->td_sampleinfo[3] != EXTRASAMPLE_UNASSALPHA) {
```

Comment:
```suggestion
                        !(td->td_extrasamples > 0 && td->td_sampleinfo[td->td_extrasamples-1] == EXTRASAMPLE_UNASSALPHA)) {
```

Reply:
trying this out right now, this doesn't seem correct after all. When converting a 4 band rgba webp to lossy jxl, I have td->td_extrasamples set to 0. ColorInterp=Alpha is correctly set on the 4th webp band.

Reply:
is PHOTOMETRIC_RGB set ? this might be a  non 8-bit image. I'm not sure the logic to automatically set PHOTOMETRIC_RGB  works currently for non 8-bit.

Reply:
td_photometric is set to PHOTOMETRIC_RGB, and the data is 8 bit